protocol/message_persistence.go

```diff
@@ -543,6 +543,91 @@ func (db sqlitePersistence) MessageByChatID(chatID string, currCursor string, li
 	return result, newCursor, nil
 }
 
+// PinnedMessageByChatID returns all pinned messages for a given chatID in descending order.
+// Ordering is accomplished using two concatenated values: ClockValue and ID.
+// These two values are also used to compose a cursor which is returned to the result.
+func (db sqlitePersistence) PinnedMessageByChatIDs(chatIDs []string, currCursor string, limit int) ([]*common.PinnedMessage, string, error) {
+	cursorWhere := ""
+	if currCursor != "" {
+		cursorWhere = "AND cursor <= ?" //nolint: goconst
+	}
+	allFields := db.tableUserMessagesAllFieldsJoin()
+	args := make([]interface{}, len(chatIDs))
+	for i, v := range chatIDs {
+		args[i] = v
+	}
+	if currCursor != "" {
+		args = append(args, currCursor)
+	}
+	// Build a new column `cursor` at the query time by having a fixed-sized clock value at the beginning
+	// concatenated with message ID. Results are sorted using this new column.
+	// This new column values can also be returned as a cursor for subsequent requests.
+	rows, err := db.db.Query(
+		fmt.Sprintf(`
+			SELECT
+				%s,
+				pm.clock_value as pinnedAt,
+				substr('0000000000000000000000000000000000000000000000000000000000000000' || m1.clock_value, -64, 64) || m1.id as cursor
+			FROM
+				pin_messages pm
+			JOIN
+				user_messages m1
+			ON
+				pm.message_id = m1.id
+			LEFT JOIN
+				user_messages m2
+			ON
+				m1.response_to = m2.id
+			LEFT JOIN
+				contacts c
+			ON
+				m1.source = c.id
+			WHERE
+				pm.pinned = 1
+				AND NOT(m1.hide) AND m1.local_chat_id IN %s %s
+			ORDER BY cursor DESC
+			LIMIT ?
+		`, allFields, "(?"+strings.Repeat(",?", len(chatIDs)-1)+")", cursorWhere),
+		append(args, limit+1)..., // take one more to figure our whether a cursor should be returned
+	)
+	if err != nil {
+		return nil, "", err
+	}
+	defer rows.Close()
+
+	var (
+		result  []*common.PinnedMessage
+		cursors []string
+	)
+	for rows.Next() {
+		var (
+			message  common.Message
+			pinnedAt uint64
+			cursor   string
+		)
+		if err := db.tableUserMessagesScanAllFields(rows, &message, &pinnedAt, &cursor); err != nil {
+			return nil, "", err
+		}
+		pinnedMessage := &common.PinnedMessage{
+			Message:  message,
+			PinnedAt: pinnedAt,
+		}
+		result = append(result, pinnedMessage)
+		cursors = append(cursors, cursor)
+	}
+
+	var newCursor string
+	if len(result) > limit {
+		newCursor = cursors[limit]
+		result = result[:limit]
+	}
+	return result, newCursor, nil
```

Comment:
The `cursors` slice maybe nil at this point, as there is no check on the slice's state.

```suggestion
	if len(result) > limit  && cursors != nil  {
		newCursor = cursors[limit]
		result = result[:limit]
	}
	return result, newCursor, nil
```